Results files aren't written during error

If I run a benchmark and an error occurs in the middle no results files are written.  

For example, say this is expected to run for 1 hour with `<time>3600</time>`, and at `00:00:00` I start:
```
java -jar benchbase.jar -b tpcc -c ....
```
Say at time `00:00:59`  I get a exception:
```
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: The connection is closed.
```

In this case, nothing at all is written to file, and I lose 59 minutes of throughput/latency/etc data for that run.  This could be useful for debugging for running experiments.

Would it be possible to change things so at least partial results files are written/flushed at the end before exiting?  Maybe some flushes or intermediate writes?  Or is the concern that this would hurt performance?

--- a/src/main/java/com/oltpbenchmark/BenchmarkState.java
+++ b/src/main/java/com/oltpbenchmark/BenchmarkState.java
@@ -92,6 +92,12 @@ public final class BenchmarkState {
         state = State.MEASURE;
     }
 
+    public void signalError() {
+        // A thread died, decrement the count and set error state
+        notDoneCount.decrementAndGet();
+        state = State.ERROR;
+    }
+
     public void startCoolDown() {
         state = State.DONE;
 

--- a/src/main/java/com/oltpbenchmark/ThreadBench.java
+++ b/src/main/java/com/oltpbenchmark/ThreadBench.java
@@ -205,8 +205,9 @@ public class ThreadBench implements Thread.UncaughtExceptionHandler {
                 }
             }
 
-            // Go to next phase if this one is complete
-            if (phaseComplete && !lastEntry) {
+            // Go to next phase if this one is complete or enter if error was thrown
+            boolean errorThrown = testState.getState() == State.ERROR;
+            if ((phaseComplete || errorThrown) && !lastEntry) {
                 // enters here after each phase of the test
                 // reset the queues so that the new phase is not affected by the
                 // queue of the previous one
@@ -355,19 +356,25 @@ public class ThreadBench implements Thread.UncaughtExceptionHandler {
 
     @Override
     public void uncaughtException(Thread t, Throwable e) {
-
-        // HERE WE HANDLE THE CASE IN WHICH ONE OF OUR WOKERTHREADS DIED
+        // Here we handle the case in which one of our worker threads died
         LOG.error(e.getMessage(), e);
-        System.exit(-1);
-
-        /*
-         * Alternatively, we could keep an HashMap<Thread,Worker> storing the
-         * runnable for each thread, so that we can get the latency numbers from
-         * a thread that died, and either continue or at least report current
-         * status. (Remember to remove this thread from the list of threads to
-         * wait for)
-         */
-
+        // We do not continue with the experiment. Instead, bypass rest of
+        // phases that were left in the test and signal error state.
+        // The rest of the workflow to finish the experiment remains the same,
+        // and partial metrics will be reported (i.e., until failure happened).
+        synchronized (testState) {
+            for (WorkloadConfiguration workConf : this.workConfs) {
+                synchronized (workConf.getWorkloadState()) {
+                    WorkloadState workState = workConf.getWorkloadState();
+                    Phase phase = workState.getCurrentPhase();
+                    while (phase != null) {
+                        workState.switchToNextPhase();
+                        phase = workState.getCurrentPhase();
+                    }
+                }
+            }
+            testState.signalError();
+        }
     }
 
     public static final class TimeBucketIterable implements Iterable<DistributionStatistics> {

--- a/src/main/java/com/oltpbenchmark/types/State.java
+++ b/src/main/java/com/oltpbenchmark/types/State.java
@@ -21,5 +21,5 @@
 package com.oltpbenchmark.types;
 
 public enum State {
-    WARMUP, MEASURE, COLD_QUERY, LATENCY_COMPLETE, DONE, EXIT,
+    WARMUP, MEASURE, COLD_QUERY, LATENCY_COMPLETE, DONE, EXIT, ERROR
 }

